pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -28,41 +30,90 @@ import (
 	"google.golang.org/grpc"
 
 	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/sets"
 	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
+	"k8s.io/kubernetes/pkg/kubelet/lifecycle"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
 )
 
+// ActivePodsFunc is a function that returns a list of pods to reconcile.
+type ActivePodsFunc func() []*v1.Pod
+
 // ManagerImpl is the structure in charge of managing Device Plugins.
 type ManagerImpl struct {
 	socketname string
 	socketdir  string
 
-	endpoints map[string]*endpoint // Key is ResourceName
+	endpoints map[string]Endpoint // Key is ResourceName
 	mutex     sync.Mutex
 
+	server *grpc.Server
+
+	// activePods is a method for listing active pods on the node
+	// so the amount of pluginResources requested by existing pods
+	// could be counted when updating allocated devices
+	activePods ActivePodsFunc
+
+	// callback is used for updating devices' states in one time call.
+	// e.g. a new device is advertised, two old devices are deleted and a running device fails.
 	callback MonitorCallback
 
-	server *grpc.Server
+	// allDevices contains all of registered resourceNames and their exported device IDs.
+	allDevices map[string]sets.String
+
+	// allocatedDevices contains allocated deviceIds, keyed by resourceName.
+	allocatedDevices map[string]sets.String
+
+	// podDevices contains pod to allocated device mapping.
+	podDevices podDevices
 }
 
 // NewManagerImpl creates a new manager on the socket `socketPath`.
-// f is the callback that is called when a device becomes unhealthy.
-// socketPath is present for testing purposes in production this is pluginapi.KubeletSocket
-func NewManagerImpl(socketPath string, f MonitorCallback) (*ManagerImpl, error) {
+// updateCapacityFunc is called to update ContainerManager capacity when device
+// capacity changes.
+func NewManagerImpl(updateCapacityFunc func(v1.ResourceList), socketPath string) (*ManagerImpl, error) {
 	glog.V(2).Infof("Creating Device Plugin manager at %s", socketPath)
 
 	if socketPath == "" || !filepath.IsAbs(socketPath) {
 		return nil, fmt.Errorf(errBadSocket+" %v", socketPath)
 	}
 
 	dir, file := filepath.Split(socketPath)
-	return &ManagerImpl{
-		endpoints: make(map[string]*endpoint),
+	manager := &ManagerImpl{
+		endpoints: make(map[string]Endpoint),
+
+		socketname:       file,
+		socketdir:        dir,
+		allDevices:       make(map[string]sets.String),
+		allocatedDevices: make(map[string]sets.String),
+		podDevices:       make(podDevices),
+	}
+
+	manager.callback = func(resourceName string, added, updated, deleted []pluginapi.Device) {
```

Comment:
prefer promoting such a function to be a method of ManagerImpl.
However, it's better be done in a follow-up PR to avoid inflating this one.

Reply:
You need to close over the `updateCapacityFunc` in the closure, so that the callback can make changes within the container manager (from endpoint). I don't like the callback - but minimizing the diff to a move, if possible.